### ISSUE TITLE
Results differ between emmeans v1.3.1 and v1.3.2

### ISSUE THREAD

**Issue by strengejacke** [rvlenth/emmeans#83]
I'm not sure if this is limited to `polr`-models, but here I found an inconsistency?

```r
library(MASS)
library(emmeans)
options(contrasts = c("contr.treatment", "contr.poly"))
fit <- polr(Sat ~ Infl + Type + Cont, weights = Freq, data = housing)

# version 1.3.2
emmeans(fit, c("Sat", "Infl"), at = list(c(Type = "Tower", Cont = "Low")), mode = "prob")
#>  Sat    Infl        prob         SE  df asymp.LCL asymp.UCL
#>  High   Low    0.2737419 0.01603723 Inf 0.2423095 0.3051743
#>  Low    Low    0.4594126 0.01912280 Inf 0.4219326 0.4968926
#>  Medium Low    0.2668455 0.01129051 Inf 0.2447165 0.2889745
#>  High   Medium 0.3939156 0.01811155 Inf 0.3584176 0.4294136
#>  Low    Medium 0.3307171 0.01737397 Inf 0.2966648 0.3647695
#>  Medium Medium 0.2753673 0.01127126 Inf 0.2532760 0.2974585
#>  High   High   0.5646865 0.02443021 Inf 0.5168041 0.6125688
#>  Low    High   0.1975453 0.01701315 Inf 0.1642001 0.2308904
#>  Medium High   0.2377683 0.01213933 Inf 0.2139756 0.2615609^

# version 1.3.1
emmeans(fit, c("Sat", "Infl"), at = list(c(Type = "Tower", Cont = "Low")), mode = "prob")
#>  Sat    Infl        prob         SE  df asymp.LCL asymp.UCL
#>  Low    Low    0.4594126 0.01912280 Inf 0.4219326 0.4968926
#>  Medium Low    0.2668455 0.01129051 Inf 0.2447165 0.2889745
#>  High   Low    0.2737419 0.01603723 Inf 0.2423095 0.3051743
#>  Low    Medium 0.3307171 0.01737397 Inf 0.2966648 0.3647695
#>  Medium Medium 0.2753673 0.01127126 Inf 0.2532760 0.2974585
#>  High   Medium 0.3939156 0.01811155 Inf 0.3584176 0.4294136
#>  Low    High   0.1975453 0.01701315 Inf 0.1642001 0.2308904
#>  Medium High   0.2377683 0.01213933 Inf 0.2139756 0.2615609
#>  High   High   0.5646865 0.02443021 Inf 0.5168041 0.6125688
```

**Comment by rvlenth** [comment]
Hmmm. What I see is the same estimate, but permuted within levels of `Infl`. I'm not sure whay that'd happen, but I'lll look. In the new version, I made a change that corrected a bug in the ordering of results for nested models, and all I can guess is that this is related. 

Do you have an idea of which result is right? I can cross my fingers and hope that I inadvertently fixed a bug in the old version. Dream on, I suppose...

**Comment by rvlenth** [comment]
Say, I just noticed that the results are the *same*. What's different are the orders of the levels of `Sat`. In your version 1.3.1 illustration, they are in ordinal order, and in the 1.3.2 version, they are alphabetical. So I think I know exactly what is happening; it is indeed related to issue #71; see the code changes in https://github.com/rvlenth/emmeans/commit/3fd30c0eef24066b4655c2a7b04cf256578552f8.  The `prob` method puts in the ordinal response as a factor, but it needs to go in as an ordered factor. I'll keep you posted.

**Comment by rvlenth** [comment]
I have fixed it and it will be updated on github the next time I push it. Thanks again for reporting this.

Since I just updated **emmeans** on CRAN, I'm reluctant to submit another update too very soon -- especially in that, technically, the results are correct, just in the wrong order. Do you think there ios urgency in making that update?

**Comment by strengejacke** [comment]
No, I don't think it's urgent. I just came accross this issue when running tests of one of my packages, where I compare the values of the first row of the returned results, which in this case broke the tests.

Is it a general problem with ordered factors, or is it just models with multinomial / categorical outcome?

**Comment by rvlenth** [comment]
I did some additional testing before committing the changes. The bug did not affect multinomial models or models with multivariate responses. You'll note that in your example even, the levels of `Infl` go in order `Low`, `Medium`, `High` even though this isn't alphabetical. The difference with `Sat` is that its levels are added via a post-`ref_grid()` call that calculates the probabilities from the cumulative probabilities. That creates a new factor in the grid and I had failed to specify its levels be in order of appearance. Other factors in a reference grid have their original definitions and that includes their correct sorting order.

In versions prior to 1.3.2, there was a bug whereby `emmeans()`'s results could be assigned to the wrong labels -- or in some cases, be completely wrong. This happened when the actual internal grid was not in the regular order needed by the algorithm that computes the marginal averages. Such re-ordering is rare, but it occurs in nested models (where it was noticed), and potentially could occur if a user uses `rbind()` or `[]` to manipulate the grid. The solution was a very simple 3 lines of code to sort the grid in standard order before computing EMMs, and that messed-up ordinal models in `prob` mode.
